datafusion/physical-optimizer/src/filter_pushdown.rs

```diff
@@ -632,6 +608,106 @@ fn push_down_filters(
     Ok(res)
 }
 
+/// A helper structure for filtering elements from a vector through multiple passes while
+/// tracking their original indices, allowing results to be mapped back to the original positions.
+struct FilteredVec<T> {
+    items: Vec<T>,
+    // Chain of index mappings: each Vec maps from current level to previous level
```

Comment:
while this multi-level mapping is clever, maybe it could be simpler if we just hard coded the number of levels needed 🤔

Reply:
it turns out a bit more complex actually because you need to keep track of 2 states (1 or 2 levels) instead of just multi-level